File: dashboard/src/lib/api/builder.ts
import { request } from './client';

// -- Types --------------------------------------------------------------------

export interface TemplateInfo {
	name: string;
	description: string;
}

export interface ModelOption {
	name: string;
	description: string;
}

export interface ProviderModels {
	provider: string;
	models: ModelOption[];
}

export interface ProviderPreset {
	name: string;
	label: string;
	base_url: string;
	api_key_env: string;
	placeholder: string;
	key_configured: boolean;
}

export interface EnvVarStatus {
	name: string;
	is_set: boolean;
}

export interface TemplateSetup {
	steps: string[];
	env_vars: EnvVarStatus[];
	extras: string[];
	docs_url: string;
}

export interface ProviderStatus {
	provider: string;
	env_var: string;
	is_configured: boolean;
}

export interface BuilderOptions {
	templates: TemplateInfo[];
	providers: ProviderModels[];
	detected_provider: string | null;
	detected_model: string | null;
	role_dirs: string[];
	custom_presets: ProviderPreset[];
	ollama_models: string[];
	ollama_base_url: string;
	template_setups: Record<string, TemplateSetup>;
	provider_status: ProviderStatus[];
	tool_func_map: Record<string, string[]>;
}

export interface StarterInfo {
	slug: string;
	name: string;
	description: string;
	tags: string[];
	features: string[];
	kind: string;
}

export interface StartersResponse {
	starters: StarterInfo[];
}

export interface ValidationIssue {
	field: string;
	message: string;
	severity: 'error' | 'warning' | 'info';
}

export interface EmbeddingOption {
	provider: string;
	env_var: string;
	is_configured: boolean;
}

export interface EmbeddingWarning {
	llm_provider: string;
	feature: string;
	current_provider: string;
	options: EmbeddingOption[];
	message: string;
}

export interface SeedResult {
	yaml_text: string;
	explanation: string;
	issues: ValidationIssue[];
	ready: boolean;
	embedding_warning: EmbeddingWarning | null;
}

export interface SaveResult {
	path: string;
	valid: boolean;
	issues: string[];
	next_steps: string[];
	agent_id: string;
}

export interface HubSearchResult {
	owner: string;
	name: string;
	description: string;
	tags: string[];
	downloads: number;
	latest_version: string;
}

export interface HubSearchResponse {
	items: HubSearchResult[];
}

// -- API functions ------------------------------------------------------------

export function getBuilderOptions(): Promise<BuilderOptions> {
	return request('/api/builder/templates');
}

export function seedAgent(body: {
	mode: 'template' | 'description' | 'blank' | 'starter';
	name: string;
	template?: string;
	description?: string;
	starter_slug?: string;
	provider: string;
	model?: string;
	base_url?: string;
	api_key_env?: string;
}): Promise<SeedResult> {
	return request('/api/builder/seed', {
		method: 'POST',
		body: JSON.stringify(body)
	});
}

export function getStarters(): Promise<StartersResponse> {
	return request('/api/builder/starters');
}

export function validateYaml(yaml_text: string): Promise<SeedResult> {
	return request('/api/builder/validate', {
		method: 'POST',
		body: JSON.stringify({ yaml_text })
	});
}

export function setEmbeddingProvider(body: {
	yaml_text: string;
	embedding_provider: string;
}): Promise<SeedResult> {
	return request('/api/builder/set-embedding-provider', {
		method: 'POST',
		body: JSON.stringify(body)
	});
}

export function saveAgent(body: {
	yaml_text: string;
	directory: string;
	filename: string;
	force?: boolean;
}): Promise<SaveResult> {
	return request('/api/builder/save', {
		method: 'POST',
		body: JSON.stringify(body)
	});
}

export function hubSearch(query: string, tags?: string[]): Promise<HubSearchResponse> {
	const params = new URLSearchParams({ q: query });
	if (tags) tags.forEach((t) => params.append('tag', t));
	return request(`/api/builder/hub-search?${params}`);
}

export function hubFeatured(): Promise<HubSearchResponse> {
	return request('/api/builder/hub-featured');
}

export function hubSeed(body: {
	ref: string;
	provider: string;
	model?: string;
	base_url?: string;
	api_key_env?: string;
}): Promise<SeedResult> {
	return request('/api/builder/hub-seed', {
		method: 'POST',
		body: JSON.stringify(body)
	});
}

File: initrunner/dashboard/schemas/builder.py
"""Agent builder, hub, and starter models."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

from initrunner.dashboard.schemas._common import (
    ProviderModels,
    ProviderPreset,
    ProviderStatus,
    ValidationIssueResponse,
)

__all__ = [
    "BuilderOptionsResponse",
    "EmbeddingOption",
    "EmbeddingWarning",
    "EnvVarStatus",
    "HubSearchResponse",
    "HubSearchResultResponse",
    "HubSeedRequest",
    "SaveRequest",
    "SaveResponse",
    "SeedRequest",
    "SeedResponse",
    "SetEmbeddingProviderRequest",
    "StarterInfo",
    "StartersResponse",
    "TemplateInfo",
    "TemplateSetup",
    "ValidateRequest",
]


class TemplateInfo(BaseModel):
    name: str
    description: str


class EnvVarStatus(BaseModel):
    name: str
    is_set: bool


class TemplateSetup(BaseModel):
    steps: list[str]
    env_vars: list[EnvVarStatus]
    extras: list[str]
    docs_url: str


class BuilderOptionsResponse(BaseModel):
    templates: list[TemplateInfo]
    providers: list[ProviderModels]
    detected_provider: str | None = None
    detected_model: str | None = None
    role_dirs: list[str]
    custom_presets: list[ProviderPreset]
    ollama_models: list[str]
    ollama_base_url: str
    template_setups: dict[str, TemplateSetup] = {}
    provider_status: list[ProviderStatus] = []
    tool_func_map: dict[str, list[str]] = {}  # tool type -> function names


class StarterInfo(BaseModel):
    slug: str
    name: str
    description: str
    tags: list[str]
    features: list[str]
    kind: str = "Agent"  # Agent, Team, or Flow


class StartersResponse(BaseModel):
    starters: list[StarterInfo]


class SeedRequest(BaseModel):
    mode: Literal["template", "description", "blank", "starter", "langchain", "pydanticai"]
    name: str
    template: str | None = None
    description: str | None = None
    starter_slug: str | None = None
    langchain_source: str | None = None
    pydanticai_source: str | None = None
    provider: str
    model: str | None = None
    base_url: str | None = None
    api_key_env: str | None = None


class EmbeddingOption(BaseModel):
    """An available embedding provider the user can choose."""

    provider: str  # "openai", "google", "ollama"
    env_var: str  # "OPENAI_API_KEY", "GOOGLE_API_KEY", "" for ollama
    is_configured: bool  # key is set / ollama running


class EmbeddingWarning(BaseModel):
    """Warning when the generated YAML needs embeddings but the effective provider is unusable."""

    llm_provider: str  # user's LLM provider (e.g. "xai")
    feature: str  # "RAG", "memory", or "RAG and memory"
    current_provider: str  # effective embedding provider
    options: list[EmbeddingOption]  # available providers with status
    message: str  # human-readable explanation


class SeedResponse(BaseModel):
    yaml_text: str
    explanation: str
    issues: list[ValidationIssueResponse]
    ready: bool
    embedding_warning: EmbeddingWarning | None = None
    sidecar_source: str | None = None
    import_warnings: list[str] = Field(default_factory=list)


class SetEmbeddingProviderRequest(BaseModel):
    yaml_text: str
    embedding_provider: str  # target provider from allowlist


class ValidateRequest(BaseModel):
    yaml_text: str


class SaveRequest(BaseModel):
    yaml_text: str
    directory: str
    filename: str
    force: bool = False
    sidecar_source: str | None = None


class SaveResponse(BaseModel):
    path: str
    valid: bool
    issues: list[str]
    next_steps: list[str]
    agent_id: str
    generated_assets: list[str] = Field(default_factory=list)


# -- InitHub -------------------------------------------------------------------


class HubSearchResultResponse(BaseModel):
    owner: str
    name: str
    description: str
    tags: list[str]
    downloads: int
    latest_version: str


class HubSearchResponse(BaseModel):
    items: list[HubSearchResultResponse]


class HubSeedRequest(BaseModel):
    ref: str  # "owner/name@version"
    provider: str
    model: str | None = None
    base_url: str | None = None
    api_key_env: str | None = None
